flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DefaultJobManagerRunnerRegistry.java

```diff
@@ -83,9 +83,16 @@ public Collection<JobManagerRunner> getJobManagerRunners() {
     }
 
     @Override
-    public CompletableFuture<Void> localCleanupAsync(JobID jobId, Executor unusedExecutor) {
+    public CompletableFuture<Void> localCleanupAsync(
+            JobID jobId, Executor ignoredExecutor, Executor mainThreadExecutor) {
         if (isRegistered(jobId)) {
-            return unregister(jobId).closeAsync();
+            CompletableFuture<Void> resultFuture = this.jobManagerRunners.get(jobId).closeAsync();
+
+            return resultFuture.thenApplyAsync(
+                    result -> {
+                        mainThreadExecutor.execute(() -> unregister(jobId));
+                        return result;
+                    });
```

Comment:
This might need to be slightly different since right now when this method is called, the result of the `closeAsync()`'s future is expected:
https://github.com/apache/flink/pull/25027/files#diff-780ffdbcb09b706ad230014f4b30772e393c67fdae253d9cf6a32adc188c5359R89-R95

Reply:
the return type is `CompletableFuture<Void>` where `Void` is just a placeholder type. There is no value in passing through the specific instance that's returned by `closeAsync` here.

Reply:
the return type is CompletableFuture<Void> where Void is just a placeholder type. There is no value in passing through the specific instance that's returned by closeAsync here.

Reply:
the return type is CompletableFuture<Void> where Void is just a placeholder type. There is no value in passing through the specific instance that's returned by closeAsync here.

Reply:
the return type is CompletableFuture<Void> where Void is just a placeholder type. There is no value in passing through the specific instance that's returned by closeAsync here.

Reply:
I'm not aware if there's functional use but the `jobId` is returned through the future and is used in assertions such as:
https://github.com/apache/flink/blob/8dc212c32036c3a4afd3af2e50e95cb87c8cee23/flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java#L715.

Reply:
I've been banging my head against this for a while now but can't figure this out. Do you have any thoughts @XComp ?

Using
```java
                    .thenRunAsync(() -> unregister(jobId), mainThreadExecutor);
```
sends the test `testNotArchivingSuspendedJobToHistoryServer` into a deadlock at line:

```java
        dispatcher.getJobTerminationFuture(jobId, Duration.ofHours(1)).join();
```


But
```java
                    .thenRunAsync(() -> mainThreadExecutor.execute(() -> unregister(jobId)));
```
doesn't.

Reply:
The Termination Future is acquired but `.join()` ends up waiting indefinitely. Also, I see the implementation of `.thenRunAsync(<task>, <executor>)` is different from just `<executor>.execute(<task>)`.

Reply:
The difference between your previous version of the code (using `thenApplyAsync`) and the changed code (using `thenRunAsync`) is a special (kind of annoying) issue with `CompletableFutures`:

The async `unregister` call is preceded by the [TestingJobManagerRunner#closeAsync](https://github.com/apache/flink/blob/153614716767d5fb8be1d4caa5e72167ce64a9d7/flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingJobManagerRunner.java#L132) which is set up to be non-blocking. That means that the method's `CompletableFuture` is completed before returning.

There's a different runtime behavior between chained operations of an already-completed future and a not-yet-completed future. In the former case, the chained operation will be executed right-away on the calling thread . In the latter case, the chained operation will be scheduled for execution.

In our scenario (see [DefaultJobManagerRunnerRegistry#localCleanupAsync](https://github.com/apache/flink/blob/e37d9c9b7b649030d9cb5d3ba6540a006aa0aa23/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DefaultJobManagerRunnerRegistry.java#L91)), the `closeAsync` call returns the already-completed future. Keep in mind that this happens on the component's main thread! So, what happens with the chained operation?
* with `thenApplyAsync`: The `mainThreadExecutor#execute` call is called asynchronously (no executor is specified, so the JVM common thread pool is used). The `unregister` callback will be scheduled on the main thread again. But `localCleanupAsync` returns the future of the `thenApplyAsync` which completes as soon as the `mainThreadExecutor#execute` returns (not the `unregister` call!). That unblocks the "buggy" code I mentioned in my previous comment.
* with `thenRunAsync`: We're on the main thread. `thenRunAsync` states that the `unregister` call should be executed on the main thread as well. But the `unregister` callback will be executed rightaway (instead of putting it on a task queue) due to the fact that the future that was returned by the previous `closeAsync` is already completed. In the mean time, the main thread is still blocked by the `ResourceCleaner` trying to complete the cleanup in the main thread. We reach a deadlock.